blog/2018/12/20/mojolicious-serving-angular-web-app/index.markdown

```diff
@@ -0,0 +1,379 @@
+
+    push @{$self->static->paths} => '/home/sachin/workspace/project/mojo_angular/mojo_angular_app/public/NgDemo';
+
+##### d. Run hypnotoad server to see if angular page is served
+
+    Sachin@02:48 PM[~/workspace/project/mojo_angular/mojo_angular_app]$ hypnotoad -f script/mojo_angular_app 
+    [Sat Dec 15 14:49:03 2018] [info] Listening at "http://*:8080"
+    Server available at http://127.0.0.1:8080
+    [Sat Dec 15 14:49:03 2018] [info] Manager 40633 started
+    [Sat Dec 15 14:49:03 2018] [info] Creating process id file "/Users/Sachin/workspace/project/mojo_angular/mojo_angular_app/script/hypnotoad.pid"
+    [Sat Dec 15 14:49:03 2018] [info] Worker 40634 started
+    [Sat Dec 15 14:49:03 2018] [info] Worker 40635 started
+    [Sat Dec 15 14:49:03 2018] [info] Worker 40637 started
+    [Sat Dec 15 14:49:03 2018] [info] Worker 40636 started
+
+![mojolicious serving angular SPA](mojo_serving_angular.png)
+
+###### Congratulations! we have served angular app with Mojolicious. Please note that the URL is the public route of mojo app `http://localhost:8080/NgDemo/`
+
+
+## Growing App
+A simple demo to show an api call to Mojolicious routes from Angular and display in Angular. 
+Since, this is not an Angular blog I will not go too deep explaining Angular; there are plenty of resources in internet for that.
+
+#### 1. Create a new route in Mojolicious App:
+Add a route `advent/2018/detail` in Mojolicious APP class(`lib/MojoAngularApp.pm`) which just responds to http `get` request 
+with first 3 Mojolicious advent 2018 detail list.
+
+    # in full mojo app it is best to put routes methods in
+    # AppName/Controller/SomeModule.pm. This is just for quick demo.
+    $r->get('advent/2018/detail' => sub {
+        my $self = shift;
+
+        return $self->render(
+            json =>
+                [
+                    {
+                        title  => 'Welcome & MojoConf Recap',
+                        day    => 1,
+                        author => 'Joel Berger',
+                    },
+                    {
+                        title  => 'Automatic Reload for Rapid Development',
+                        day    => 2,
+                        author => 'Doug Bell',
+                    },
+                    {
+                        title  => 'Higher Order Promises',
+                        day    => 3,
+                        author => 'brain d foy',
+                    },
+                ],
+            status => 200,
+        );
+    });
+
+You may have to allow Cross Origin Resource Sharing(CORS) if Angular app throws an error while accessing the mojo API endpoint.
+In that case you may want to make use of awesome `before_dispatch` app hook like bellow:
+
+    $self->hook(before_dispatch => sub {
+        my $c = shift;
+        $c->res->headers->header('Access-Control-Allow-Origin' => '*');
+    });
+
+Certainly, `*` is too insecure you might just give `localhost`. This is demo and I am lazy to experiment.
```

Comment:
I wouldn't recommend them to use localhost either. Just mention that in a real-world application `*` is defeating the security feature.